guides/v2.2/coding-standards/technical-guidelines.md

```diff
@@ -532,7 +532,75 @@ class View extends Template
 
 ### 6.4. Service Contracts (Application) layer
 
-We are reviewing this section and will publish it soon.
+6.4.1. Location
+
+6.4.1.1. Service contract interfaces should be placed in separate API modules. All other modules will depend on the API module, while implementations could be easily swapped via `di.xml`. For example, if module is named `MyModule`, its APIs should be declared in the module named `MyModuleApi`.
+
+6.4.1.2. Service interfaces which should be exposed as web APIs must be placed under `MyModuleApi/Api/Service` directory. Service data interfaces must be placed under `MyModuleApi/Api/Data`. 
+
+6.4.1.3. All other APIs, including explicit extension points like Chain or Composite implementations must be placed under `MyModuleApi/Model`. 
+
+6.4.1.4. Arbitrary directory nesting can be used for APIs to improve code structure. For example, `MyCompany\MyModuleApi\Api\Service\Relations\DeleteInterface`.
+
+6.4.2. Types Hierarchy
+
+6.4.2.1. If service data interface should be extensible, it must extend from `Magento\Framework\Api\ExtensibleDataInterface`.
+
+6.4.2.1. Extensible data interfaces must not form hierarchies. Extension attributes do not work for child extensible interfaces.
+
+6.4.3. Service Interface Structure
+
+6.4.3.1. Methods named similarly must serve similar purpose across different services, but still might have different signatures.
+
+6.4.3.2. Services must only operate on scalar types or service data interfaces.
+
+6.4.3.3. Service contracts should not be used for read scenarios on the storefront, GraphQL should be used for storefront scenarios instead. Check out [web API technical vision]({{ page.baseurl }}/coding-standards/technical-vision/webapi.html) for more details.
+
+6.4.3.4. Each service interface should declare a single public method. Interface name should reflect task/action to be performed. Example `Magento\InventoryApi\Api\StockSourceLinksDeleteInterface::execute(array $links)`. The only exception is Repository API which may be added for convenience and must be limited to singular CRUD operations and `getList($searchCriteria)`.
+
+6.4.4. Service Method Signature
+
+6.4.4.1. Service contracts should not apply presentation layer formatting to the returned data.
+
+6.4.4.2. Service data interfaces represent domain entities and can reference related data interfaces.
+
+6.4.4.3. Service data interfaces must be implemented as DTOs with a set of symmetric getters and setters.
+
+6.4.4.3. Strict typing is enforced for Service and Data interfaces located under `MyCompany/MyModuleApi/Api`. Only the following types are allowed:
+
+* Scalar types: `string` (including Date and DateTime); `int`; `float`; `boolean`
```

Comment:
I would suggest to create and use our own types for date types. We could have API similar to what Java has, Date, DateTime, LocalDate, LocalDateTime. I can look into this.

Reply:
this sounds like a more sophisticated topic, as you are proposing to introduce Value-Objects into Magento.

What distinguishes Value Object from Entity:
- the absence of ID or UUID
- the absence of setter methods (Value Objects are immutable only)

Even if we would introduce such entities, not sure that we should distinguish Value Objects from Data Interfaces (entities) in this document.

Reply:
I think this can be clarified separately in the future in scope of independent PR.